modules/packages/DistributedDeque.chpl

```diff
@@ -59,638 +57,712 @@ use Collection;
-    if localThis.cap > 0 {
-      while true {
-        var size = queueSize.fetchAdd(1);
+      // We find our slot based on another fetch-add counter, making this wait-free as well.
+      var head = globalHead.fetchAdd(1) % nSlots;
+      var elt = slots[abs(head)].popFront();
+      var retval = (true, elt);
+      return retval;
+    }
 
-        // Over capacity, fix before returning...
-        if size >= localThis.cap {
-          on this {
-            var readSize = queueSize.read();
-            // Attempt to fix, but yield to reduce potential contention and CPU hogging.
-            while readSize > this.cap && !queueSize.compareExchangeWeak(readSize, this.cap) {
-              chpl_task_yield();
-              readSize = queueSize.read();
-            }
-          }
+    /*
+      Obtains the number of elements held by this queue.
+    */
+    proc getSize() : int {
+      return queueSize.read();
+    }
 
-          // Fixed, return early...
-          return false;
-        }
-        // At this point, we either have a capacity and something between [0, cap],
-        // or we do not have a capacity and we have something greater than 0,
-        // either of which means we are safe.
-        else if size >= 0 {
-          break;
+    /*
+      Performs a lookup for the element in the data structure.
+    */
+    proc contains(elt : eltType) : bool {
+      var containsElem : atomic bool;
+      forall elem in this {
+        if elem == elt {
+          containsElem.write(true);
         }
-        // If the size is negative, a dequeuer has overcommitted. Since there can
-        // be an element for us, we loop again.
       }
+      return containsElem.read();
     }
-    // If we have a 'negative' capacity, we do not have one and can hold an
-    // infinite amount; this gives us a more optimized code path.
-    else {
-      while true {
-        var size = queueSize.fetchAdd(1);
 
-        // We have reserved a spot for our operation.
-        if size >= 0 {
-          break;
+    /*
+      Iterate over all elements in the deque in the order specified.
+      
+      
+
+      **FIXME:** Likely can be worked around by either using snapshot iteration approach
+      or by making the lock reentrant and forcing all iterators to acquire all locks in
+      some global locking order.
+
+      **FIXME:** Ordered serial iterators currently are not working well when the head or tail are negative
+      (because the head and tail may need to move backwards, its easily possible for it to be negative. We
+      handle this right now with the core methods, but iteration wasn't the focus during development...), 
+      but there is not enough time to fix it. This is an official warning against using anything but 
+      iteration when `order` == `Ordering.NONE` (A.K.A the default unordered iteration). 
+
+      **TODO:** Need to create a parallel iterator (not really parallel, just allowed to be used in a

```

Comment:
I think parallel unordered iteration *can* be allowed. If a user uses a deque as an iterand in a `forall` I think we can assume that they want parallel iteration and by nature, it is unordered. Similarly reduction can be unordered, as well.

Reply:
By default, parallel iteration *is* unordered. My complaint is that there is no way to reduce it in a way that is ordered.

"Why would it need to be ordered?" - Imagine the following...

```chpl
var deque = new DistDeque(string);
deque.pushBack("World");
deque.pushFront("Hello");
writeln(+ reduce deque);
```

This will not be in any ordering at all. Reduction isn't only useful for summation, sometimes ordering can matter, I would like to the user to be able to do `+ reduce deque.these(ordering.FIFO)` or if thats not possible syntactically:

`deque.reduceFIFO(lambda(s1 : string, s2 : string) : string { return s1 + s2; });`

That would be pretty cool overall, and very useful.

Reply:
I think my main concern in the comment was "not really parallel" part. And my question is "why not?"

In terms of ordering in reduction:

First, you cannot use an iterator of your choosing, nor pass an argument to `these` while doing reduction. It simply is not in the grammar.

Second, the use case you are presenting is not a very realistic use case for reductions. And I can't think of one, myself. I think reduction operations should always be commutative, so ordering is not necessary. Even if this wasn't the case, I think this is a bigger question than your implementation of the parallel iterator, and parallel iterator should still be unordered and *parallel*.

What you are showing here needs to be achieved by a scan operation and not by reduction. And scan operations currently use serial iterators anyways.

Reply:
I suppose I was a bit unclear, but what I meant in my message is that we need *another* parallel iterator, that wasn't parallel and just yielded sequentially in order. However, I realize now that this isn't needed, as reductions are supported if you declare the iterator to use, like [this](https://tio.run/##dVHBaoRADD3rVzy8qFT2A1Y8Fhb2UtrCHkoPwYl1IOvIaGtL6bfbGbWuPXgJk7z3kpdJWVPLMo5hKdR10D3b@0@6tsL4DgOfoq@54yT1eVAZi7KGbhCdWMTgYqyoCMrgS7MoB@Zh8BMu0kpXxivxT/rXcCvCRjXjLVm6oqc3HCdbZ92oFEPNlqdqUazlgzApJ/QOP8iCrEWBl/gUZ4jZB1mD8QE@XNbUrqiKX/PZK4ksdn23nf32nWaojPuf4bnW3Z7rmbH4ntvfRNMYN8jvo3u3TsPD9jxJmofhYF1FmiR6Kqk5Ispwh849HdHBN5StJsEjq/eSF5qdEkc8zEfa8h/Ir897Ckcdx18).

As well, in terms of reduction, all languages I have seen so far have reductions be ordered on an ordered data structure. For example, see Haskell's [fold](https://wiki.haskell.org/Fold). I was merely suggesting providing a means to reduce in order. `scan` is nice and all, but it has an O(N) space overhead, especially when you only need the last value produced.

Reply:
Following this, its unfortunately I can't just do it this way, but if I were to provide this (big 'if' at this point), I would likely do so in parallel, and assume that all values are associative. Like, for example, imagine if I emulated a block distribution instead of cyclic, then I would have `maxTaskPar` adjacent `LocalDeque` that maintain a relative ordering...

```
LD1      LD2    LD3       LD4
-H1-    -H2-   -H3-     -H4-
-H5-    -H6-   -H7-     -H8-
-H9-    -H10- -H11-   -H12-
```

Where 'LD#' is a `LocalDeque` index, and `H#` is the index of each element from the Head. (This is how ordering is preserved). Reduction can take advantage of this by reducing each 'row' from each 'column', producing sub-results which get combined with sub-results from other nodes, etc. It remains ordered, and it is still parallel.